### DISCUSSION TITLE
A chat plugin within Kimai

### DISCUSSION THREAD

**Discussion by rarvin** [kimai/kimai#1626] [General]
Kimai is now a complete time tracking system, but Tasks has taken kimai to the next level. What becomes critical now is real time communication with the other users accessing the system. Hence having a chat system built into kimai would be a killer feature since all users are already on the system and having seamless communication becomes very critical in getting required task completed on time. For this a unified chat system in kimai would be very helpful.

**Comment by kevinpapst** [comment]
Thanks for posting that idea!
I was thinking of this for a longer time now.
The main reasons why I haven't touched this by now (besides the problem that time is rare) is that chat functionality can reach from simple one-to-one channels to ultimately complex systems.

Can you edit your initial post and add some ideas what you personally would see necessary for a first version? I mean: we all have used chat systems and know what features exists, but I just want to know the most basic functionality that you would see in this time-tracking / work assignment scenario.

**Comment by HeinzWuert** [comment]
From my own experience, sorry, but cannot agree to this need.
I have already installed company chat systems several times, but in the end you have to say very clearly:
- Chats within the LAN / in-house are accepted and well used
- Chats with external / mobile users are absolutely not.
The reason is quite simply messenger systems like Whatsapp, Telegram etc. These are kill applications for local systems because everybody is using them. Why should a user switch to a second or third power eating online-system?
For me, there are more important issues in the area of actual time recording than this.
My opinion!

**Comment by kevinpapst** [comment]
I agree that there are other important issues with time-tracking. But I do see the some good use-cases for internal notifications. Eg. when a team-lead can flag entries that needs descriptions or short communications regarding tasks... 

Don't misunderstand internal "chat" as messenger system. But as a basic communication tool for work / time-tracking / task management. Users could send it as email, but then it is not in the same place.

**Comment by HeinzWuert** [comment]
In order to be able to send an email with an automated reference to the team and / or admin about a project or a time booking, such an info could be quite practical.
Exactly like for the admins to point out a booking error to a user, for example.
This would simplify work processes.
That would not be a chat system, but an information system at the mail level.

**Comment by kevinpapst** [comment]
It is not only about pushing emails to users. But emails should be configurable as well.

**Comment by rarvin** [comment]
@kevinpapst The idea that I have felt would help is a group chat system where you have all users who are linked to the various group through the use of teams. Any one who posts a chat meesage posts the message to the group where all users who are within the team can see such messages. Also if there is a real need to indicate a specific person then @mention works where it is targetted to the specific person but all of the other users in the team are able to see that very same message.

User from other teams more so a Team lead should be able to post onto the other groups with a similar method of mention but to the entire group @group-mention. If a user in that group or team responds back to the team lead with a @mention then the message should reach the team lead.

Each user should individually be able to see all @mention messages against their name and @mention messages in the group should be visible within the whole team. If an @mention is done to a user within a group then it should be visible to all members in that group.

Within group all messages visible to all users even @mention within the group.
@mention alone visible to the individual users
any @mentiongroup to be pushed to the whole users of the said group and visible in the group.
Any response from that to a group to be via @mentiongroups.
Also team leads can be @mentioned from different groups and team by the individual users.  

This way teams can work quickly and efficiently, say if there was a task that was pending one team could ask the other team to expedite the work, members within the team could pass messages within a group. Members from one group can pass messages to another group. Members of one group can contact team leads from other groups. Individual use of mentions helps each user see all actions that the person needs to work on.

This for inter-coordinated working that chat can help. Simple chat functionality would be good and if needed maybe screenshots, and links of kimai pages or kimai task/timesheet ID's links alone which chat users can directly click and open and nothing more.

**Comment by adrianrudnik** [comment]
That sounds oddly specific like something you could link to existing solutions like Microsoft Teams / Slack / Chanty through bots and command addons. That would also inverse the responsibility for the chat and allow for features like:

- On timeline action notify team lead in "slack".
- Start/stop time through slash-chat-commands in "slack".
- Mention time entries in "slack" with a card, preview and link to kimai.
- Trigger exports through "slack".

Slack being just an example here. Could also be telegram, or mattermost or any other self hosted solution that supports some kind of integration.

**Comment by kevinpapst** [comment]
Cross linking the web hooks feature request #1407

**Comment by sideeffect42** [comment]
@rarvin: Would [JSXC](https://www.jsxc.org) maybe fit your needs? It describes itself as being easily integratable into other products, seems to be actively maintained and is based on the standard XMPP protocol.
